src/mavrosflight/sensors/mag.cpp

```diff
@@ -0,0 +1,402 @@
+      if (measurement != measurement_prev_)
+      {
+        measurements_.push_back(measurement);
+      }
+
+      measurement_throttle_ = 0;
+      measurement_prev_ = measurement;
+    }
+    measurement_throttle_++;
+  }
+  else if (calibrating_)
+  {
+    // fit ellipsoid to the measurements according to Li paper but in a RANSAC form
+    ROS_INFO("Collected %lu measurements. Fitting ellipsoid.", measurements_.size());
+    Eigen::MatrixXd u = ellipsoidRANSAC(measurements_, ransac_iters_, inlier_thresh_);
+
+    // magnetometer calibration parameters according to Renaudin paper
+    ROS_INFO("Computing calibration parameters.");
+    magCal(u, A_, b_);
+
+    calibrating_ = false;
+  }
+
+  return !calibrating_;
+}
+
+bool Mag::correct(mavlink_small_mag_t msg, double *xmag, double *ymag, double *zmag)
+{
+  *xmag = (double)msg.xmag;
+  *ymag = (double)msg.ymag;
+  *zmag = (double)msg.zmag;
+
+  return true;
+}
+
+Eigen::MatrixXd Mag::ellipsoidRANSAC(std::deque<Eigen::Vector3d> meas, int iters, double inlier_thresh)
+{
+    // initialize random number generator
+    std::random_device random_dev;
+    std::mt19937 generator(random_dev());
+    
+    // RANSAC to find a good ellipsoid fit
+    int inlier_count_best = 0; // number of inliers for best fit
+    std::deque<Eigen::Vector3d> inliers_best; // container for inliers to best fit
+    double dist_sum = 0; // sum distances of all measurements from ellipsoid surface
+    int dist_count = 0; // count number distances of all measurements from ellipsoid surface
+    for (unsigned i = 0; i < iters; i++)
+    {
+        // pick 9 random, unique measurements by shuffling measurements
+        // and grabbing the first 9
+        std::shuffle(meas.begin(), meas.end(), generator);
+        std::deque<Eigen::Vector3d> meas_sample;
+        for (unsigned j = 0; j < 9; j++)
+        {
+            meas_sample.push_back(meas[j]);
+        }
+
+        // fit ellipsoid to 9 random points
+        Eigen::MatrixXd u = ellipsoidLS(meas_sample);
+
+        // check if LS fit is actually an ellipsoid (paragraph of Li following eq. 2-4)
+        // unpack needed parts of u
+        double a = u(0);
+        double b = u(1);
+        double c = u(2);
+        double f = u(3);
+        double g = u(4);
+        double h = u(5);
+        double p = u(6);
+        double q = u(7);
+        double r = u(8);
+        double d = u(9);
+
+        double I = a + b + c;
+        double J = a * b + b * c + a *c - f * f - g * g - h * h;
+        
+        if (4 * J - I * I <= 0)
+        {
+            continue;
+        }
+
+        // eq. 15 of Renaudin and eqs. 1 and 4 of Li
+        Eigen::MatrixXd Q(3, 3);
+        Q << a, h, g,
+             h, b, f,
+             g, f, c;
+
+        Eigen::MatrixXd ub(3, 1);
+        ub << 2 * p,
+              2 * q,
+              2 * r;
+        double k = d;
+
+        // eq. 21 of Renaudin (should be negative according to eq. 16)
```

Comment:
Dude, This is awesome!  Thanks for putting in references to papers.

Any chance you could put the full reference in somewhere?